File: mpframework/common/tags.py
#--- Mesa Platform, Copyright 2021 Vueocity, LLC
"""
    Logic for matching text tags with each other.
    MPF uses as the basis to configure flexible connections
    between systems entities, like content and licenses.

    This logic is duplicated here for Python and DB matching, and
    DUPLICATED IN CLIENT CODE for optimizing comparisons there.

    Regular expression matching is supported, but since the logic
    needs to be supported for DB searches, the scope of regex
    will be limited to the DB support.
"""
import re
from django.db.models import Q
from django.core.validators import RegexValidator
from django.conf import settings

from . import log


# Django validator for content item tags
validate_item_tag = RegexValidator( re.compile( r'^[-\.\w]+\Z', re.U ),
    u"Please use only use letters, numbers, or _-.", 'invalid' )

NEGATION_DELIM = '!'
REGEX_DELIM = '/'


def tag_match( tag, tag_matches ):
    """
    Determine if a given tag matches tag_matches (case insensitive).
    DOES NOT include check for includes all, as that is managed
    separately by the APA.
    """
    rv = False
    tag = str( tag ).strip()
    if tag:
        tag = tag.lower()
        # Exact match
        rv = tag in tag_matches
        # Otherwise check case and wildcard matching
        if not rv:
            for tm in tag_matches:
                rv = _tag_match( tm, tag )
                if rv:
                    break
    return rv

def _tag_match( match_tag, item_tag ):
    # EXACT MATCH ALREADY HANDLED

    # Setup for one return with negation
    rv = False
    negative = False
    if match_tag.startswith( NEGATION_DELIM ):
        negative = True
        match_tag = match_tag.strip( NEGATION_DELIM )

    # Regular expression match
    if match_tag.startswith( REGEX_DELIM ):
        try:
            match_tag = match_tag.strip( REGEX_DELIM )
            match_re = re.compile( match_tag, re.IGNORECASE )
            rv = match_re.match( item_tag )
        except Exception as e:
            log.info("CONFIG - Bad regex tag match: %s -> %s", match_tag, e)
            if settings.MP_DEV_EXCEPTION:
                raise

    if not rv:
        # Embedded wildcard by checking for fragments between
        # wildcards in the order they are presented by removing
        # each match as it occurs
        match_text = match_tag.strip('*')
        frags = match_text.split('*')
        if len( frags ) > 1:
            rv = True
            for frag in frags:
                if frag in item_tag:
                    item_tag = item_tag.replace( frag, '', 1 )
                else:
                    rv = False
                    break
        # Check wildcards on the ends
        elif match_tag.startswith('*') and match_tag.endswith('*'):
            rv = match_text in item_tag
        elif match_tag.startswith('*'):
            rv = item_tag.endswith( match_text )
        elif match_tag.endswith('*'):
            rv = item_tag.startswith( match_text )

    return bool(rv) ^ negative


def tags_Q( tags, tag_field='tag' ):
    """
    Return Q for whether any tag match expressions in tags match
    any content tags in a query filter.
    """
    q = Q()
    for tag in tags:
        q |= tag_Q( tag, tag_field )
    return q

def tag_Q( tag, tag_field='tag' ):
    """
    Implement same logic as wildcards in tag_match, by building up
    Q queries for filter args.
    """
    q = Q()

    # Setup for negation
    negative = False
    if tag.startswith( NEGATION_DELIM ):
        negative = True
        tag = tag.strip( NEGATION_DELIM )

    # Handle regular expression
    if tag.startswith( REGEX_DELIM ):
        tag = tag.strip( REGEX_DELIM )
        q = Q(**{ tag_field + '__iregex': tag })

    if not q:
        tag_text = tag.strip('*')

        # Add cases for matches in middle
        frags = tag_text.split('*')
        if len( frags ) > 1:
            # HACK - don't try to replicate the order of fragments
            for frag in frags:
                filt = { tag_field + '__icontains': frag }
                q &= Q( **filt )

        # Handle other cases
        if not q:
            if tag.startswith('*') and tag.endswith('*'):
                filt = '__icontains'
            elif tag.startswith('*'):
                filt = '__iendswith'
            elif tag.endswith('*'):
                filt = '__istartswith'
            else:
                filt = '__iexact'
            q = Q(**{ tag_field + filt: tag_text })

    return ~q if negative else q

File: mpframework/frontend/portal/static/mpf-js/portal/tags.js
/*--- Mesa Platform, Copyright 2021 Vueocity, LLC

    Tag case-insensitive and wildcard matching logic

    Recreate most of the content tag matching logic from the
    server, to provide UI display using info available on the client.
*/
(function() { 'use strict';

    const NEGATION_DELIM = '!'
    const REGEX_DELIM = '/'

    mpp.tag_match_item = function( item, tags, nodes ) {
    /*
        Return whether any tags matches the item model's tag
        Optionally check any items in nodes (e.g., root collections).
    */
        let match = false
        const item_tag = item.get('tag') ? item.get('tag').toLowerCase() : ''
        if( tags ) {
            tags = (tags + '').split(',')
            // Check item first
            if( item_tag ) {
                _.some( tags, function( tag ) {
                    match = _tag_match( tag, item_tag )
                    return match
                    })
                }
            // Check any trees the item is part of
            if( !match && nodes ) {
                _.some( nodes, function( node ) {
                    match = mpp.tag_match_item( node, tags )
                    return match
                    })
                }
            }
        return match
        }

    function _tag_match( tag, item_tag ) {
    /*
        Replicate mpframework.common.tags matching logic.
        Duplicating logic isn't ideal, but this is easiest way.
    */
        if( !tag ) {
            return false
            }
        // Check exact match
        tag = tag.toLowerCase()
        if( tag == item_tag ) {
            return true
            }

        // Setup for one return with negation
        let rv = false
        let negative = false
        if( _.startsWith( tag, NEGATION_DELIM ) ) {
            negative = true
            tag = _.trim( tag, NEGATION_DELIM )
            }

        // Regular expression
        if( _.startsWith( tag, REGEX_DELIM ) ) {
            try {
                rv = RegExp( _.replace( tag, REGEX_DELIM, '' ) ).test( item_tag )
            }
            catch( e ) {
                mp.log_error("Bad regex tag match: ", tag, " -> ", e)
                }
            }

        // Wildcards
        if( !rv ) {
            const text = _.trim( tag, '*' )
            const frags = text.split('*')
            // Wildcards in the middle
            if( frags.length > 1 ) {
                rv = frags.every( function( frag ) {
                    if( _.includes( item_tag, frag ) ) {
                        item_tag = _.replace( item_tag, frag, '' )
                        return true
                        }
                    })
                }
            // Wildcards on the ends
            else if( _.startsWith( tag, '*' ) && _.endsWith( tag, '*' ) ) {
                rv = _.includes( item_tag, text )
                }
            else if( _.startsWith( tag, '*' ) ) {
                rv = _.endsWith( item_tag, text )
                }
            else if( _.endsWith( tag, '*' ) ) {
                rv = _.startsWith( item_tag, text )
                }
            }

        return !!rv != negative
        }

    })();
